File: 2021/day_2/day_2.py
from re import U
from typing import Match


with open('2021/day_2/day_2.txt') as file:
    data = file.read()

A = [line for line in data.split('\n')]
A = [line.split() for line in A]

def f(arr):
	n = len(arr)
	depth = 0
	horizontal = 0

	for instruction in arr:
		command, unit = instruction
		unit = int(unit)

		if command == "forward":
			horizontal += unit
		elif command == "down":
			depth += unit
		elif command == "up":
			depth -= unit

	return horizontal * depth


def f2(arr):
	n = len(arr)
	depth = horizontal = aim = 0
	for instruction in arr:
		command, unit = instruction
		unit = int(unit)

		if command == "forward":
			horizontal += unit
			depth += aim * unit
		elif command == "down":
			aim += unit
		elif command == "up":
			aim -= unit

	return horizontal * depth

print(f2(A))

File: 2021/day_2/day_2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"bufio"
	"strings"
);

func main() {
	path := "./day_2.txt"
	arr := readLines(path)

	fmt.Println("--- Part One ---")
	fmt.Println(f(arr))

	fmt.Println("--- Part Two ---")
	fmt.Println(f2(arr))
}


func f(arr []string) int {
	depth, horizontal := 0, 0
	for _, instruction := range arr {
		ins := strings.Split(instruction, " ")
		command := ins[0]
		unit := toInt(ins[1])

		switch command {
		case "forward":
			horizontal += unit
			break
		case "down":
			depth += unit
		case "up":
			depth -= unit
		default:
			// empty
		}

	}
	return depth * horizontal
}


func f2(arr []string) int {
	depth, horizontal, aim := 0, 0, 0
	for _, instruction := range arr {
		ins := strings.Split(instruction, " ")
		command := ins[0]
		unit := toInt(ins[1])

		switch command {
		case "forward":
			horizontal += unit
			depth += aim * unit
		case "down":
			aim += unit
		case "up":
			aim -= unit
		default:
			// empty
		}

	}
	return depth * horizontal
}


func readLines(filename string) []string {
	file, err := os.Open(filename)
	check(err)
	defer file.Close()

	Scanner := bufio.NewScanner(file)

	var lines []string
	for Scanner.Scan() {
		lines = append(lines, Scanner.Text())
	}
	return lines
}


func check(err error) {
	if err != nil {
		panic(err)
	}
}


func toInt(str string) int {
	res, err := strconv.Atoi(str)
	check(err)
	return res
}